File: bcv_api.js
#!/usr/bin/env node
/**
 * BCV API - Web scraping for Venezuelan Central Bank exchange rates
 * Retrieves exchange rates for USD only (Optimized)
 */

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const axios = require('axios');
const cheerio = require('cheerio');
const fs = require('fs');
const path = require('path');
// const https = require('https'); // Not used with env var approach

class BCVScraper {
    constructor() {
        this.baseUrl = 'https://www.bcv.org.ve/';
        this.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        };
        this.cacheFile = path.join(__dirname, 'bcv_rate_cache.json');
        this.cacheTime = 3600 * 1000; // 1 hour in milliseconds
    }
    async getExchangeRates() {
        // ... cache check ...
        try {
            const response = await axios.get(this.baseUrl, {
                headers: this.headers,
                timeout: 10000,
                decompress: true // Enable GZIP automatically
            });

            const $ = cheerio.load(response.data);
            const rates = this._parseRates($);

            const data = {
                status: 'success',
                timestamp: new Date().toISOString(),
                source: 'BCV (Banco Central de Venezuela)',
                rates: rates
            };

            // Save to cache
            fs.writeFileSync(this.cacheFile, JSON.stringify(data));

            return data;

        } catch (error) {
            // Try to return stale cache on error
            if (fs.existsSync(this.cacheFile)) {
                try {
                    const cachedData = JSON.parse(fs.readFileSync(this.cacheFile, 'utf8'));
                    if (cachedData) {
                        cachedData.cached = true;
                        cachedData.stale = true;
                        return cachedData;
                    }
                } catch (e) {
                    // Ignore
                }
            }

            return {
                status: 'error',
                message: `Failed to fetch data from BCV: ${error.message}`,
                timestamp: new Date().toISOString()
            };
        }
    }

    /**
     * Parse exchange rates from the HTML
     * @param {CheerioStatic} $ - Cheerio instance with loaded HTML
     * @returns {Object} Parsed exchange rates
     */
    _parseRates($) {
        const rates = {};

        // Try to find exchange rates in tables using optimized traversal
        const tables = $('table');

        for (let i = 0; i < tables.length; i++) {
            const rows = $(tables[i]).find('tr');

            for (let j = 0; j < rows.length; j++) {
                const cells = $(rows[j]).find('td, th');

                if (cells.length > 0) {
                    const currencyText = $(cells[0]).text().trim();
                    console.log(`Row[${i}][${j}]: ${currencyText} (cells: ${cells.length})`);
                }

                if (cells.length >= 2) {
                    const currencyText = $(cells[0]).text().trim();

                    // Optimized: Check directly for Dólar
                    if (currencyText.includes('Dólar') || currencyText.includes('USD')) {
                        try {
                            const rateText = $(cells[1]).text().trim();
                            // console.log(`Found USD rate text: ${rateText}`); // Debug
                            const rate = this._cleanRate(rateText);

                            if (rate !== null) {
                                rates['USD'] = {
                                    currency: 'USD',
                                    name: 'Dólar',
                                    rate: rate,
                                    symbol: '$'
                                };
                                // Found USD, break completely
                                return rates;
                            }
                        } catch (error) {
                            continue;
                        }
                    }
                }
            }
        }

        // If USD not found, add placeholder
        if (!rates['USD']) {
            rates['USD'] = {
                currency: 'USD',
                name: 'Dólar',
                rate: null,
                symbol: '$',
                note: 'Rate not available - check BCV website'
            };
        }

        return rates;
    }

    /**
     * Clean and convert rate text to float
     * @param {string} rateText - Raw rate text
     * @returns {number|null} Cleaned rate or null
     */
    _cleanRate(rateText) {
        try {
            // Remove known separators first
            let cleaned = rateText.replace(/,/g, '.').replace(/\s/g, '');
            // Strict cleanup
            cleaned = cleaned.replace(/[^\d.-]/g, '');

            if (!cleaned || isNaN(cleaned)) return null;

            return parseFloat(cleaned);
        } catch (error) {
            return null;
        }
    }
}

/**
 * Main function to run the scraper
 */
async function main() {
    const scraper = new BCVScraper();
    const data = await scraper.getExchangeRates();
    console.log(JSON.stringify(data, null, 2));
}

// Export for module usage
module.exports = BCVScraper;

// Run if executed directly
if (require.main === module) {
    main().catch(console.error);
}

File: bcv_api.py
#!/usr/bin/env python3
"""
BCV API - Web scraping for Venezuelan Central Bank exchange rates
Retrieves exchange rates for USD, EUR, Yuan, Turkish Lira, and Russian Ruble
"""

import requests
from bs4 import BeautifulSoup
import json
from datetime import datetime
from typing import Dict, Optional


class BCVScraper:
    """Scraper for BCV (Banco Central de Venezuela) exchange rates"""
    
    def __init__(self) -> None:
        self.base_url = "https://www.bcv.org.ve/"
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        }
        self.session = requests.Session()
        self.session.headers.update(self.headers)
        
    def get_exchange_rates(self) -> Dict:
        """
        Scrape exchange rates from BCV website
        
        Returns:
            dict: Exchange rates in JSON format
        """
        try:
            response = self.session.get(self.base_url, timeout=10)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, 'html.parser')
            
            # Parse exchange rates from the BCV website
            rates = self._parse_rates(soup)
            
            return {
                "status": "success",
                "timestamp": datetime.now().isoformat(),
                "source": "BCV (Banco Central de Venezuela)",
                "rates": rates
            }
            
        except requests.RequestException as e:
            return {
                "status": "error",
                "message": f"Failed to fetch data from BCV: {str(e)}",
                "timestamp": datetime.now().isoformat()
            }
        except Exception as e:
            return {
                "status": "error",
                "message": f"Error processing data: {str(e)}",
                "timestamp": datetime.now().isoformat()
            }
    
    def _parse_rates(self, soup: BeautifulSoup) -> Dict:
        """
        Parse exchange rates from the HTML soup
        
        Args:
            soup: BeautifulSoup object with the page content
            
        Returns:
            dict: Parsed exchange rates
        """
        rates = {}
        
        # Try to find exchange rates in the common BCV structure
        # Currently implemented: look for tables with exchange rate data
        tables = soup.find_all('table')
        for table in tables:
            rows = table.find_all('tr')
            for row in rows:
                cells = row.find_all(['td', 'th'])
                if len(cells) >= 2:
                    currency_text = cells[0].get_text(strip=True)
                    # Directly check for Dólar
                    if 'Dólar' in currency_text:
                        try:
                            rate_text = cells[1].get_text(strip=True)
                            # Clean and convert the rate
                            rate = self._clean_rate(rate_text)
                            if rate:
                                rates['USD'] = {
                                    "currency": 'USD',
                                    "name": 'Dólar',
                                    "rate": rate,
                                    "symbol": '$'
                                }
                                # Found what we needed, return immediately or break
                                return rates
                        except (ValueError, IndexError):
                            continue
        
        # If USD not found, add placeholder
        if 'USD' not in rates:
            rates['USD'] = {
                "currency": 'USD',
                "name": 'Dólar',
                "rate": None,
                "symbol": '$',
                "note": "Rate not available - check BCV website"
            }
        
        return rates
    
    def _clean_rate(self, rate_text: str) -> Optional[float]:
        """Clean and convert rate text to float"""
        try:
            # Remove currency symbols, spaces, and convert comma to dot
            cleaned = rate_text.replace(',', '.').replace(' ', '')
            # Remove any non-numeric characters except dot and minus
            cleaned = ''.join(c for c in cleaned if c.isdigit() or c in '.-')

            # If nothing remains after cleaning, treat as no rate
            if not cleaned:
                return None

            # Validate minus sign usage: at most one, and only at the beginning
            if cleaned.count('-') > 1:
                return None
            if '-' in cleaned and not cleaned.startswith('-'):
                return None

            # Validate decimal point usage: at most one
            if cleaned.count('.') > 1:
                return None

            # Ensure there is at least one digit
            if not any(ch.isdigit() for ch in cleaned):
                return None

            return float(cleaned)
        except ValueError:
            return None
    
    def _get_currency_symbol(self, code: str) -> str:
        """Get currency symbol for a given currency code"""
        symbols = {
            'USD': '$',
            'EUR': '€',
            'CNY': '¥',
            'TRY': '₺',
            'RUB': '₽'
        }
        return symbols.get(code, '')


def main():
    """Main function to run the scraper"""
    scraper = BCVScraper()
    data = scraper.get_exchange_rates()
    print(json.dumps(data, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()
